File: wear/src/main/java/james/metronome/MetronomeService.java
package james.metronome;

import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.media.AudioAttributes;
import android.media.AudioManager;
import android.media.SoundPool;
import android.os.Binder;
import android.os.Build;
import android.os.Handler;
import android.os.IBinder;
import android.os.VibrationEffect;
import android.os.Vibrator;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.v4.app.NotificationCompat;

public class MetronomeService extends Service implements Runnable {

    public static final String ACTION_PAUSE = "james.metronome.ACTION_PAUSE";

    public static final String PREF_VIBRATION = "vibration";
    public static final String PREF_INTERVAL = "interval";

    private final IBinder binder = new LocalBinder();

    private SharedPreferences prefs;
    private int bpm;
    private long interval;

    private SoundPool soundPool;
    private Handler handler;
    private int soundId = -1;
    private boolean isPlaying;
    private boolean isVibration;

    private Vibrator vibrator;

    private TickListener listener;

    @Override
    public void onCreate() {
        super.onCreate();
        prefs = PreferenceManager.getDefaultSharedPreferences(this);

        vibrator = (Vibrator) getSystemService(Context.VIBRATOR_SERVICE);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            soundPool = new SoundPool.Builder()
                    .setMaxStreams(1)
                    .setAudioAttributes(new AudioAttributes.Builder()
                            .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                            .build())
                    .build();
        } else soundPool = new SoundPool(1, AudioManager.STREAM_MUSIC, 0);

        isVibration = prefs.getBoolean(PREF_VIBRATION, true);
        if (!isVibration)
            soundId = soundPool.load(this, R.raw.click, 1);

        interval = prefs.getLong(PREF_INTERVAL, 500);
        bpm = toBpm(interval);

        handler = new Handler();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null && intent.getAction() != null) {
            switch (intent.getAction()) {
                case ACTION_PAUSE:
                    pause();
            }
        }
        return START_STICKY;
    }

    private static int toBpm(long interval) {
        return (int) (60000 / interval);
    }

    private static long toInterval(int bpm) {
        return (long) 60000 / bpm;
    }

    public void play() {
        handler.post(this);
        isPlaying = true;

        Intent intent = new Intent(this, MetronomeService.class);
        intent.setAction(ACTION_PAUSE);

        startForeground(530,
                new NotificationCompat.Builder(this)
                        .setContentTitle(getString(R.string.notification_title))
                        .setContentText(getString(R.string.notification_desc))
                        .setSmallIcon(R.drawable.ic_notification)
                        .setContentIntent(PendingIntent.getService(this, 0, intent, PendingIntent.FLAG_ONE_SHOT))
                        .build()
        );

        if (listener != null)
            listener.onStartTicks();
    }

    public void pause() {
        handler.removeCallbacks(this);
        stopForeground(true);
        isPlaying = false;

        if (listener != null)
            listener.onStopTicks();
    }

    public void setBpm(int bpm) {
        this.bpm = bpm;
        interval = toInterval(bpm);
        prefs.edit().putLong(PREF_INTERVAL, interval).apply();
    }

    public void setVibration(boolean vibration) {
        isVibration = vibration;
        if (!isVibration)
            soundId = soundPool.load(this, R.raw.click, 1);
        else soundId = -1;

        prefs.edit().putBoolean(PREF_VIBRATION, isVibration).apply();
    }

    public boolean isPlaying() {
        return isPlaying;
    }

    public long getInterval() {
        return interval;
    }

    public int getBpm() {
        return bpm;
    }

    public boolean isVibration() {
        return isVibration;
    }

    public void setTickListener(TickListener listener) {
        this.listener = listener;
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return binder;
    }

    @Override
    public boolean onUnbind(Intent intent) {
        listener = null;
        return super.onUnbind(intent);
    }

    @Override
    public void onDestroy() {
        handler.removeCallbacks(this);
        super.onDestroy();
    }

    @Override
    public void run() {
        if (isPlaying) {
            if (soundId != -1)
                soundPool.play(soundId, 1.0f, 1.0f, 0, 0, 1.0f);
            else if (Build.VERSION.SDK_INT >= 26)
                vibrator.vibrate(VibrationEffect.createOneShot(50, VibrationEffect.DEFAULT_AMPLITUDE));
            else vibrator.vibrate(50);

            handler.postDelayed(this, interval);
        }
    }

    public class LocalBinder extends Binder {
        public MetronomeService getService() {
            return MetronomeService.this;
        }
    }

    public interface TickListener {
        void onStartTicks();

        void onStopTicks();
    }
}

File: app/src/main/java/me/jfenn/metronome/services/MetronomeService.kt
package me.jfenn.metronome.services

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.media.AudioAttributes
import android.media.AudioManager
import android.media.SoundPool
import android.os.*
import android.preference.PreferenceManager
import androidx.core.app.NotificationCompat
import me.jfenn.metronome.R
import me.jfenn.metronome.utils.PREF_EMPHASIS
import me.jfenn.metronome.utils.PREF_INTERVAL
import me.jfenn.metronome.utils.PREF_TICK
import me.jfenn.metronome.utils.PreferenceDelegate
import me.jfenn.metronome.views.TicksView
import java.util.*

class MetronomeService : Service() {

    private val binder: IBinder = LocalBinder()
    private val prefs: SharedPreferences by lazy { PreferenceManager.getDefaultSharedPreferences(this) }

    var interval: Long by PreferenceDelegate<Long>(PREF_INTERVAL, 500) {
        listener?.onBpmChanged(toBpm(it))
    }

    // this.bpm just delegates itself to `this.interval`
    var bpm: Int
        get() = toBpm(interval)
        set(value) { interval = toInterval(value) }

    private var timer = Timer()
    private var timerTask = MetronomeTask(this)

    fun createTimerTask(delay: Long, interval: Long) {
        val oldTimer = timer
        timerTask = MetronomeTask(this)
        timer = Timer().apply {
            scheduleAtFixedRate(timerTask, delay, interval)
        }

        // cancel the old timer to stop executions (should be gc'd shortly after)
        oldTimer.cancel()
    }

    fun cancelTimerTask() {
        timer.cancel()
    }

    var tick: Int by PreferenceDelegate(PREF_TICK, 0) {
        soundId = createSoundId(tick)
    }

    var isPlaying = false
    private var emphasisIndex = 0
    var emphasisList: MutableList<Boolean> by PreferenceDelegate(PREF_EMPHASIS, mutableListOf(false, false, false, false))

    private var soundId: Int = -1
    private val soundPool: SoundPool by lazy {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            SoundPool.Builder()
                    .setMaxStreams(1)
                    .setAudioAttributes(AudioAttributes.Builder()
                            .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                            .build())
                    .build()
        } else SoundPool(1, AudioManager.STREAM_MUSIC, 0)
    }

    private val vibratorService: Vibrator by lazy { getSystemService(Context.VIBRATOR_SERVICE) as Vibrator }
    private val notificationManager: NotificationManager by lazy { getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager }

    private var listener: TickListener? = null

    override fun onCreate() {
        super.onCreate()
        soundId = createSoundId(tick)
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        when (intent?.action) {
            ACTION_START -> {
                bpm = intent.getIntExtra(EXTRA_BPM, bpm)
                pause()
                play()
            }
            ACTION_PAUSE -> pause()
        }

        return START_STICKY
    }

    override fun onBind(intent: Intent): IBinder? {
        return binder
    }

    override fun onUnbind(intent: Intent): Boolean {
        listener = null
        return super.onUnbind(intent)
    }

    override fun onDestroy() {
        cancelTimerTask()
        super.onDestroy()
    }

    fun setTickListener(listener: TickListener?) {
        this.listener = listener
    }

    fun play() {
        isPlaying = true
        emphasisIndex = 0
        val intent = Intent(this, MetronomeService::class.java).apply { action = ACTION_PAUSE }

        val builder: NotificationCompat.Builder = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            notificationManager.createNotificationChannel(NotificationChannel("metronome", getString(R.string.app_name), NotificationManager.IMPORTANCE_LOW).apply {
                setSound(null, null)
                enableVibration(false)
            })

            NotificationCompat.Builder(this, "metronome")
        } else NotificationCompat.Builder(this)

        startForeground(530,
                builder.setContentTitle(getString(R.string.notification_title))
                        .setContentText(getString(R.string.notification_desc))
                        .setSmallIcon(R.drawable.ic_notification)
                        .setContentIntent(PendingIntent.getService(this, 0, intent, PendingIntent.FLAG_ONE_SHOT))
                        .setPriority(NotificationCompat.PRIORITY_LOW)
                        .setSound(null)
                        .setVibrate(longArrayOf(0L))
                        .build()
        )

        createTimerTask(0, interval)
        listener?.onStartTicks()
    }

    fun pause() {
        cancelTimerTask()
        stopForeground(true)

        isPlaying = false
        listener?.onStopTicks()
    }

    private fun createSoundId(tick: Int): Int {
        return if (tick >= 0 && tick < TicksView.ticks.size && !TicksView.ticks[tick].isVibration)
            soundPool.load(this, TicksView.ticks[tick].soundRes, 1)
        else -1
    }

    inner class LocalBinder : Binder() {
        val service: MetronomeService
            get() = this@MetronomeService
    }

    private inner class MetronomeTask(
            val service: MetronomeService
    ) : TimerTask() {

        var interval = -1L
        val mainHandler = Handler(Looper.getMainLooper())

        override fun run() {
            if (service.isPlaying) {
                if (interval == -1L)
                    interval = service.interval

                if (service.emphasisIndex >= service.emphasisList.size)
                    service.emphasisIndex = 0

                val isEmphasis = service.emphasisList[service.emphasisIndex]

                when {
                    service.soundId != -1 -> service.soundPool.play(service.soundId, 1f, 1f, 0, 0, if (isEmphasis) 1.5f else 1f)
                    Build.VERSION.SDK_INT >= 26 -> service.vibratorService.vibrate(VibrationEffect.createOneShot(15L, if (isEmphasis) 255 else 70))
                    else -> service.vibratorService.vibrate(if (isEmphasis) 50L else 20L)
                }

                mainHandler.post {
                    service.listener?.onTick(isEmphasis, service.emphasisIndex)
                    service.emphasisIndex++
                }

                if (interval != service.interval) {
                    interval = service.interval
                    service.createTimerTask(interval, interval)
                }
            }
        }
    }

    interface TickListener {
        fun onStartTicks()
        fun onTick(isEmphasis: Boolean, index: Int)
        fun onBpmChanged(bpm: Int)
        fun onStopTicks()
    }

    companion object {
        const val ACTION_START = "james.metronome.ACTION_START"
        const val ACTION_PAUSE = "james.metronome.ACTION_PAUSE"
        const val EXTRA_BPM = "james.metronome.EXTRA_BPM"

        private fun toBpm(interval: Long): Int {
            return (60000 / interval).toInt()
        }

        private fun toInterval(bpm: Int): Long {
            return 60000.toLong() / bpm
        }
    }
}